be/src/service/service_be/lake_service.cpp

```diff
@@ -937,6 +938,124 @@ void LakeServiceImpl::compact(::google::protobuf::RpcController* controller, con
+
+    // 2. check status
+    if (cntl->Failed()) {
+        ac_context->handle_failure("link rpc channel failed");
+        return;
+    } else {
+        if (response->status().status_code() != 0) {
+            ac_context->handle_failure("call compact failed");
+            return;
+        }
+    }
+
+    // 3. collect txn logs
+    ac_context->collect_txnlogs(response);
+}
+
+void LakeServiceImpl::aggregate_compact(::google::protobuf::RpcController* controller,
+                                        const ::starrocks::AggregateCompactRequest* request,
+                                        ::starrocks::CompactResponse* response, ::google::protobuf::Closure* done) {
+    brpc::ClosureGuard guard(done);
+    auto cntl = static_cast<brpc::Controller*>(controller);
+
+    if (request->requests_size() == 0) {
+        cntl->SetFailed("empty requests");
+        return;
+    }
+    if (request->compute_nodes_size() != request->requests_size()) {
+        cntl->SetFailed("compute nodes size not equal to requests size");
+        return;
+    }
+
+    AggregateCompactContext ac_context;
+    ac_context.latch = std::make_unique<BThreadCountDownLatch>(request->requests_size());
+
+    for (int i = 0; i < request->requests_size(); i++) {
+        const auto& single_req = request->requests(i);
+        const auto& compute_node = request->compute_nodes(i);
+        if (!compute_node.has_host() || !compute_node.has_brpc_port()) {
+            ac_context.handle_failure("compute node missing host/port");
+            ac_context.count_down();
+            continue;
+        }
+        brpc::Controller* node_cntl = new brpc::Controller();
+        CompactResponse* node_resp = new CompactResponse();
+        node_cntl->set_timeout_ms(single_req.timeout_ms());
+        butil::EndPoint endpoint;
+        std::string brpc_url = fmt::format("{}:{}", compute_node.host(), compute_node.brpc_port());
+        if (str2endpoint(brpc_url.c_str(), &endpoint)) {
+            ac_context.handle_failure("unknown endpoint, host=" + compute_node.host());
+            ac_context.count_down();
+            continue;
+        }
+        std::unique_ptr<brpc::Channel> channel(new brpc::Channel());
+        brpc::ChannelOptions options;
+        options.connect_timeout_ms = config::rpc_connect_timeout_ms;
+        channel->Init(endpoint, &options);
+        // TODO stub cache
+        auto stub = std::make_shared<starrocks::LakeService_Stub>(channel.release(),
+                                                                  google::protobuf::Service::STUB_OWNS_CHANNEL);
+        stub->compact(node_cntl, &single_req, node_resp,
+                      brpc::NewCallback(aggregate_compact_cb, node_cntl, node_resp, &ac_context));
+    }
+
+    // wait for all tasks to finish
+    ac_context.wait();
+
+    // write combined txn log
+    // TODO // submit write_combined_txn_log to thread pool to avoid block brpc thread.
+    ac_context.write_combined_txn_log();
```

Comment:
shall check the status explicitly before call the write_combined_txn_log or the interface returns a Status to indicate whether the write is good or not.

Current `void write_combined_txn_log()` gives the impression that the txn log will be always written and never fails by justing reading the interface declaration.

Reply:
I think we already handle failure internally.

Reply:
We wrap the failure handle

Reply:
done